File: apps/api-fsharp/src/SalesManagement/Domain/LotWorkflows.fs
module SalesManagement.Domain.LotWorkflows

open SalesManagement.Domain.Types

type ManufacturingCompletionError = LotNotInManufacturing
type ShippingInstructionError = LotNotManufactured
type ShippingCompletionError = LotNotShippingInstructed
type CancellationError = LotNotManufactured
type ItemConversionError = LotNotManufacturedForConversion
type ItemConversionCancelError = LotNotInConversionInstructed

let completeManufacturing (date: System.DateOnly) (lot: ManufacturingLot) : ManufacturedLot =
    { Common = lot.Common
      ManufacturingCompletedDate = date }

let instructShipping (deadline: System.DateOnly) (lot: ManufacturedLot) : ShippingInstructedLot =
    { Common = lot.Common
      ManufacturingCompletedDate = lot.ManufacturingCompletedDate
      ShippingDeadlineDate = deadline }

let completeShipping (date: System.DateOnly) (lot: ShippingInstructedLot) : ShippedLot =
    { Common = lot.Common
      ManufacturingCompletedDate = lot.ManufacturingCompletedDate
      ShippingDeadlineDate = lot.ShippingDeadlineDate
      ShippedDate = date }

let cancelManufacturingCompletion (lot: ManufacturedLot) : ManufacturingLot = { Common = lot.Common }

let instructItemConversion (info: ConversionDestinationInfo) (lot: ManufacturedLot) : ConversionInstructedLot =
    { Common = lot.Common
      ManufacturingCompletedDate = lot.ManufacturingCompletedDate
      DestinationInfo = info }

let cancelItemConversionInstruction (lot: ConversionInstructedLot) : ManufacturedLot =
    { Common = lot.Common
      ManufacturingCompletedDate = lot.ManufacturingCompletedDate }

File: apps/api-spring/domain/src/main/java/com/example/salesmanagement/domain/LotWorkflows.java
package com.example.salesmanagement.domain;

import java.time.LocalDate;

public final class LotWorkflows {
    private LotWorkflows() {}

    public static ManufacturedLot completeManufacturing(LocalDate date, ManufacturingLot lot) {
        return new ManufacturedLot(lot.common(), date);
    }

    public static ShippingInstructedLot instructShipping(LocalDate deadline, ManufacturedLot lot) {
        return new ShippingInstructedLot(lot.common(), lot.manufacturingCompletedDate(), deadline);
    }

    public static ShippedLot completeShipping(LocalDate date, ShippingInstructedLot lot) {
        return new ShippedLot(lot.common(), lot.manufacturingCompletedDate(), lot.shippingDeadlineDate(), date);
    }

    public static ManufacturingLot cancelManufacturingCompletion(ManufacturedLot lot) {
        return new ManufacturingLot(lot.common());
    }

    public static ConversionInstructedLot instructItemConversion(
            ConversionDestinationInfo destination, ManufacturedLot lot) {
        return new ConversionInstructedLot(lot.common(), lot.manufacturingCompletedDate(), destination);
    }

    public static ManufacturedLot cancelItemConversionInstruction(ConversionInstructedLot lot) {
        return new ManufacturedLot(lot.common(), lot.manufacturingCompletedDate());
    }
}
